weave/trace_server/trace_server_common.py

```diff
@@ -1,6 +1,90 @@
 import copy
+import datetime
+import json
 from collections import OrderedDict
-from typing import Any, Dict, Optional
+from typing import Any, Dict, Literal, Optional, Union, cast
+
+from weave.trace_server import trace_server_interface as tsi
+from weave.trace_server.refs_internal import InternalObjectRef, parse_internal_uri
+
+
+def make_derived_summary_fields(
+    call_dict: Dict[str, Any], summary_key: Literal["summary", "summary_dump"]
+) -> tsi.SummaryMap:
+    """
+    Make derived summary fields for a call.
+
+    Summary is controlled by the user, but the `weave` summary key is
+    used to store derived fields, adhering to the tsi.SummaryMap type.
+
+    Args:
+        call_dict: The call dict.
+        summary_key: The key in the call dict that contains the summary. In
+            the clickhouse server this is "summary_dump", but in sqlite
+            it's "summary".
+
+    Returns:
+        The derived summary fields.
+    """
+    display_name = call_dict.get("display_name")
+    started_at_dt = _make_datetime_from_any(call_dict["started_at"])
+    ended_at_dt = _make_datetime_from_any(call_dict.get("ended_at"))
+
+    status = "success"
+    if call_dict.get("exception"):
+        status = "error"
+    elif ended_at_dt is None:
+        status = "running"
+
+    latency = None
+    if ended_at_dt and started_at_dt:
+        latency = (ended_at_dt - started_at_dt).microseconds
+
+    if not display_name:
+        op = parse_internal_uri(call_dict["op_name"])
+        if isinstance(op, InternalObjectRef):
+            display_name = op.name
+        else:
+            display_name = call_dict["op_name"]
+
+    summary = _load_json_maybe(call_dict.get(summary_key)) or {}
+
+    weave_summary = summary.get("weave", {})
+    weave_summary["trace_name"] = display_name
+    weave_summary["status"] = status
+    if latency is not None:
+        weave_summary["latency"] = latency
+
+    summary["weave"] = weave_summary
+    return cast(tsi.SummaryMap, summary)
+
+
+def _make_datetime_from_any(
+    dt: Optional[Union[str, datetime.datetime]],
+) -> Optional[datetime.datetime]:
+    """
+    Flexible datetime parser, accepts None, str and datetime.
+    This allows database type agnostic parsing of dates.
+    """
+    if dt is None:
+        return None
+    if isinstance(dt, str):
+        return datetime.datetime.fromisoformat(dt)
+    elif isinstance(dt, datetime.datetime):
+        return dt
+    return None
+
+
+def _load_json_maybe(value: Any) -> Any:
```

Comment:
same question here - why are we sometimes strings and sometimes dicts?

Reply:
yo we have two different databases, in clickhouse these bad boys are json, in sqlite they need loading

Reply:
we could massage the data before

Reply:
oh i see. i think i would pass a flag in top of the main function saying something like `data_is_stringified` or something like that. That the callers pass in. Then perform the proper paths from there. The reason i say this is i want to avoid cases where we accidentally have the wrong stored/in-mem format somewhere upstream, and this function happily consumes it.

A bit of a nit. you could also just comment this in the those functions.

Reply:
removed